packages/compass-generative-ai/src/tools-controller.ts

```diff
@@ -47,12 +148,161 @@ export class ToolsController {
+    // this.connectionManager. The mongodb-mcp-server does not seem to deal with
+    // a state where the connection is still being established by the time the
+    // tool executes, so it won't wait for us. In that situation the tool call
+    // will just fail.
+    void this.switchConnectionManagerToCurrentConnection(context.connection);
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-redundant-type-constituents
+  get server(): Server | undefined {
+    return this.runner?.server;
+  }
+
+  public async startServer(): Promise<void> {
+    if (this.runner.server) {
+      return;
+    }
+
+    try {
+      this.logger.log.info(
+        this.logger.mongoLogId(1_001_000_402),
+        'ToolsController',
+        'Starting in-memory MCP server with config',
+        {
+          ...this.runner.userConfig,
+          apiClientId: '<redacted>',
+          apiClientSecret: '<redacted>',
+        }
+      );
+
+      await this.runner.start();
+
+      this.logger.log.info(
+        this.logger.mongoLogId(1_001_000_403),
+        'ToolsController',
+        'In-memory MCP server started successfully'
+      );
+    } catch (error) {
+      // In case of errors we don't want Compass to crash so we
+      // silence MCP start errors and instead log them for debugging.
+      this.logger.log.error(
+        this.logger.mongoLogId(1_001_000_404),
+        'ToolsController',
+        'Error when attempting to start MCP server',
+        { error }
+      );
+    }
+  }
+
+  public async stopServer(): Promise<void> {
+    try {
+      if (!this.runner.server) {
+        this.logger.log.info(
+          this.logger.mongoLogId(1_001_000_406),
+          'ToolsController',
+          'MCP server stop requested. No MCP server running, nothing to stop.'
+        );
+        return;
+      }
+      await this.runner.server.close();
+      this.logger.log.info(
+        this.logger.mongoLogId(1_001_000_407),
+        'ToolsController',
+        'MCP server stopped successfully'
+      );
+    } catch (error) {
+      this.logger.log.error(
+        this.logger.mongoLogId(1_001_000_408),
+        'ToolsController',
+        'Error when attempting to close the MCP server',
+        { error }
+      );
+    }
+  }
+
+  private async switchConnectionManagerToCurrentConnection(
+    connectParams?: ToolsConnectParams
+  ): Promise<void> {
+    try {
+      this.currentConnection = connectParams;
+
+      await this.connectionManager.updateConnection(connectParams);
+    } catch (error) {
+      this.logger.log.error(
+        this.logger.mongoLogId(1_001_000_410),
+        'ToolsController',
+        'Error when attempting to switch connection for connection manager',
+        { error }
+      );
+    }
+  }
+}
+
+// TODO: this is a complete hack because server's registerTools() is private and
+// the only way it gets called is through connect() which is never going to
+// happen.
+function registerTools(server: Server) {
```

Comment:
The mongodb-mcp-server branch used by the PoC had a getAvailableTools() method on the server object which nicely sidestepped all this, but alas that doesn't exist in main or the latest release. I think we can avoid this pretty easily with a small change to mongodb-mcp-server.